### DISCUSSION TITLE
How to use Custom Validators with 1:N relationships?

### DISCUSSION THREAD

**Discussion by jwbensley** [netbox-community/netbox#9992] [Q&A]
I am using custom validators. They are working fine for basic checks.

I am now trying to add some custom validation to VRFs. If I do a simple `print(vars(instance))` to see what is passed to the `validate()` function in my custom validator class, I see the following:

```
        {
            '_state': <django.db.models.base.ModelState object at 0x7f59cb0da6b0>,
            'id': 1,
            'created': datetime.datetime(2022, 8, 9, 13, 15, 38, 144593, tzinfo=datetime.timezone.utc),
            'last_updated': datetime.datetime(2022, 8, 9, 15, 56, 36, 662701, tzinfo=datetime.timezone.utc),
            'custom_field_data': {
                'vrf_type': 'IP Transit'
            },
            'name': 'DFZ',
            'rd': '2',
            'tenant_id': 1,
            'enforce_unique': True,
            'description': '',
            '_prechange_snapshot': {
                'created': '2022-08-09T13:15:38.144Z',
                'last_updated': '2022-08-09T15:56:36.662Z',
                'name': 'DFZ',
                'rd': '2',
                'tenant': 1,
                'enforce_unique': True,
                'description': '',
                'import_targets': [1, 2, 3, 4],
                'export_targets': [1],
                'custom_fields': {
                    'vrf_type': 'IP Transit'
                },
                'tags': []
            }
        }
```

In the above output, I am editing an existing VRF, but the same applies when creating a new one; the `import_targets` and `export_targets` values aren't present in the instance. They are only present under `_prechange_snapshot`, which is only present if you are editing an existing VRF, so in the case you are adding a new VRF, `_prechange_snapshot` isn't there at all. Even so, `_prechange_snapshot` is showing you what they were, not what they are now, so in either case (new or existing VRF), the `instance` object passed to the validator function is missing the two route target lists.

How can one validate which RTs are being added/edited against a VRF?

**Comment by jwbensley** [comment]
This also happens with the [BGP plugin](https://github.com/k01ek/netbox-bgp) but it's nothing to do with the plugin. When adding a BGP sessions I am unable to validate the import/export policies. These object attributes, like the route target attributes in my VRF example above, are simply missing from the object passed to the custom validator function.

I believe this is happening because just as a VRF has a 1:N relationship with RTs, BGP sessions have a 1:N relationship with import/export policies. In this case when an obj has a 1:N relationship, the list of those foreign objects aren't passed to the custom validator function.

What I think is happening is that NetBox creates the VRF, or BGP session, or whatever, and then makes the relationships to the foreign objects (RTs in the case of VRFs of import/export policies in the case of the BGP session), as a second step.

Is it possible that somehow we can get these objects passed to the obj bassed to the validator function?

**Reply by larsux** [reply]
I had the same issue with tags. They are also not passed to the custom validator function, so that you are not able to do validation based on updated tags.

That leads me to the same Question: Is it possible to access related objects (i.e. tags) from the "transient update request" (so not from the stored Instance via _prechange_snapshot) in the custom validator?

**Comment by BarbarossaTM** [comment]
Just hit the same wall, when trying to validate semantic tags.

I've build

```
from extras.validators import CustomValidator

import re

SEMANTIC_TAG_RE = re.compile(r"^([^ ]+)=(.+)$")

class SemanticTagValidator(CustomValidator):
    def validate(self, instance):
        # Object doesn't support tags
        if not hasattr(instance, "tags"):
            return

        # Object doesn't have tags assigned
        tags = list(instance.tags.names())
        if len(tags) == 0:
            return

        keys = {}
        for tag in instance.tags.names():
            match = SEMANTIC_TAG_RE.search(str(tag))
            if not match:
                continue

            key = match.group(1)
            if keys.get(key):
                self.fail(f"Can't have two semantic tags with the same key '{key}'", field='tags')

            keys[key] = True
```

and applied it to `dcim.interfaces` and other things. I'm seeing the same behavior that the validation only happens against what tags are already in the DB and not the ones being entered :grimacing:
